crates/wasi/src/preview2/udp.rs

```diff
@@ -42,6 +43,9 @@ pub struct UdpSocket {
 
     /// Socket address family.
     pub(crate) family: SocketAddressFamily,
+
+    /// The pool of allowed address
+    pub(crate) pool: Arc<Pool>,
```

Comment:
Could this start as `None` to represent that it's bound-late? That could also help overwriting it by accident by only writing to it if it's `None`

Reply:
(same below too)

Reply:
Pushed a change that does this, but I have a question about the exact desired semantics of `start-bind`. Take the following_example:

```rust
// `bad_addr` is an address that is not permitted
assert!(sock.start_bind(bad_addr).is_err());
// `good_addr` is permitted
sock.start_bind(good_addr);
```

As currently written the second call to `start-bind` will fail with `InvalidState` since the pool has already been set on the socket. Is this right? Previously, this would have succeeded. I would imagine that `start_bind` should be allowed to be called multiple times as long as the socket has not successfully been bound before.

Reply:
I removed the check that the pool was not previously set as this broke some tests. Once we decide on what the semantics should be, I can adjust.

Reply:
cc @badeend on this you are probably best-equipped to answer this

Reply:
@rylev 

From the API's perspective, unbound sockets (`Default`) never need a pool, and all other states (`BindStarted`, `Bound`, `Connected`) could've only be reached by calling `bind` at some point, so they always have access to a pool. A way to codify this could be:

```rs
pub(crate) enum UdpState {
    Default,
    BindStarted(Arc<Pool>),
    Bound(Arc<Pool>),
    Connected(Arc<Pool>),
}
```

Also, OutgoingDatagramStream's `pool` field doesn't have to be optional, because you can never call `stream` on an unbound socket.

---

> As currently written the second call to start-bind will fail with InvalidState since the pool has already been set on the socket. Is this right?

From the caller's perspective, the `start-bind` call should preferably succeed or fail atomically. Using the enum layout above would _automagically_ fix that.